### DISCUSSION TITLE
在NVIDIA 的GPU机器上执行eglInitialize()函数卡死

### DISCUSSION THREAD

**Discussion by cgeffect** [Tencent/libpag#2446] [Q&A] [answered]
之前在没有GPU的机器上, 使用swiftshader作为后端渲染, 没有问题, 但是现在把代码迁移到NVIDIA的GPU机器上, 源代码没有改动, 工程链接的libEGL.so 和 libGLESv2.so 也没有改动, 使用的是在没有GPU的机器上编译出来的. 运行时发现在执行
```
  static std::shared_ptr<PAGSurface> MakeOffscreen(int width, int height);
```
时卡死, 代码跟到里面最终定位在EGLGlobals.cpp的
```
  eglInitialize(globals.display, &majorVersion, &minorVersion);
```
上一步执行的  globals.display = eglGetDisplay(EGL_DEFAULT_DISPLAY); 返回的globals.display值是1.

GPU的信息
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.54.03              Driver Version: 535.54.03    CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A30                     On  | 00000000:65:01.0 Off |                    0 |
| N/A   33C    P0              33W / 165W |  12724MiB / 24576MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A     29696      C   /opt/tritonserver/bin/tritonserver        11224MiB |
|    0   N/A  N/A     29703      C   python3                                     242MiB |
|    0   N/A  N/A     30065      C   /usr/bin/python3                            242MiB |
|    0   N/A  N/A     30621      C   /usr/bin/python3                            242MiB |
|    0   N/A  N/A     30622      C   /usr/bin/python3                            242MiB |
|    0   N/A  N/A     30623      C   /usr/bin/python3                            242MiB |
|    0   N/A  N/A     30624      C   /usr/bin/python3                            242MiB |
+---------------------------------------------------------------------------------------+

**Comment by kevingpqi123** [comment, marked as answer]
pag 在 NVIDIA 的 GPU 机器渲染没有问题 ，使用的前提是 GPU 的渲染环境需要跑通，这里都是标准的 OpenGL ES 渲染，环境的问题需要业务方自己解决，可以基于 Mesa 构建 EGL 渲染环境。

**Reply by cgeffect** [reply]
请问下, 只需要把EGL环境跑通就可以了吗?代码和编译pag有需要改动的地方吗?

**Reply by kevingpqi123** [reply]
印象中好像不需要，先确保基于 EGL 可以正常渲染， 还有就是含有 BMP 预合成的 pag 文件需要视频解码，如果想要性能好，需要去适配 GPU 硬件解码

**Reply by cgeffect** [reply]
已解决, 确实是EGL环境的问题.
